python/ray/tests/test_basic.py

```diff
@@ -2801,3 +2801,32 @@ def test_redis_lru_with_set(ray_start_object_store_memory):
 
     # Now evict the object from the object store.
     ray.put(x)  # This should not crash.
+
+
+@pytest.mark.parametrize("two_node_cluster", [{"num_cpus": 10}], indirect=True)
+def test_call_shutdown_in_remote_function(two_node_cluster):
+    num_cpus = 10
+
+    @ray.remote
+    class Actor(object):
+        def shutdown(self):
+            time.sleep(1)
+            ray.shutdown()
```

Comment:
@guoyuhong I think we should disallow this and raise an exception if someone tries to call `ray.shutdown()` outside of the driver process. What do you want the behavior to be in this case? I didn't understand your explanation in https://github.com/ray-project/ray/issues/4495#issuecomment-477473027.

I think instead in this case the actor should just call `sys.exit()` or maybe `self.__ray_terminate__()`

Reply:
@robertnishihara `sys.exit()` won't work, because the backend will treat this as an unintentional exit and reconstruct the actor.

`self.__ray_terminate__` can work, but I think it's less preferable. One minor reason is that `__ray_terminate__` is embedded to the class by Ray, not written by the user. So it's kind of weird for the user to call `self.__ray_terminate__`. Moreover, another major reason is that Java doesn't allow programmatically embedding a method into a class. So we cannot have such a method, say `this.rayTerminate`, in Java. 

Thus, I think maybe we can change `shutdown` to something like this:

```
def shutdown(intentional_exit=False):
   """
   Disconnect this driver or worker from Raylet.

  Args:
     intentional_exit (bool): This can only be set to True in actor. When set to True, the current 
        actor will be treated as intentionally exited, and won't be reconstructed.
   """  
```

Reply:
Alternatively, another solution is to have an extra `ray.exit_actor` method. This might be more clear than calling `ray.shutdown(intentional_exit=True)` in actors.

Reply:
I view `ray.shutdown()` as doing the opposite of `ray.init()`. `ray.init()` starts all of the Ray processes, and `ray.shutdown()` terminates all of the processes. So using `ray.shutdown()` to disconnect an actor from the cluster doesn't make sense to me.

I agree that it's weird for the user to call `__ray_terminate__`, but I also think it's weird to need to do this manually. Why not just call `del` on the relevant actor handle?

Reply:
@robertnishihara agreed that `ray.shutdown` should be the opposite of `ray.init`. After thinking more about this, I also feel that `ray.shutdown(intentionally_exit=True)` isn't nice. Because it makes the `shutdown` method too complicated. Users have to do think about the combination of the arguments. 

Regarding deleting the actor handle, the problem is that currently we don't have reference counting for the forked handles. (Actually we had to disable `ActorHandle.__del__` to work around this bug.)

What do you think about the second solution, `ray.exit_actor`?

Reply:
Are you saying that you always shutdown actors manually?

What's wrong with just keeping the original actor handle in scope for the lifetime of the actor? Is it too hard to keep it in scope? Or is it too hard to know when it's no longer being used?

How would the actor be able to tell when it should call `ray.exit_actor`?

Would `ray.exit_actor` just call the actor's `__ray_terminate__` method?

Reply:
For long running actors, we do not necessarily need to call the shutdown.

However, it is different to some other cases. We use an actor to do some model serving and a new model is coming, we will start a new actor with a new port. After the new actor is up, we then need shutdown the old actor to release the port and memory occupied by it.

Reply:
> Are you saying that you always shutdown actors manually?

Yes, the current GC mechanism doesn't work for us.

> What's wrong with just keeping the original actor handle in scope for the lifetime of the actor?

For a long-running job, we exit the driver right after it pulls up the actors (otherwise the driver becomes the single point of failure).  Thus, the original handle is garbage collected when the driver exits.

> How would the actor be able to tell when it should call ray.exit_actor?

Users need to manually shutdown an actor when they think the actor isn't needed any more.

> Would ray.exit_actor just call the actor's __ray_terminate__ method?

I'm thinking of the other way around. E.g.

```
def exit_actor():
   global_worker.raylet_client.disconnect()
   sys.exit(0)

def __ray_terminate__(self):
   ray.exit_actor()
```

Reply:
@guoyuhong @raulchen you said the issue with just calling `sys.exit()` is that Ray might try to reconstruct the actor. However, if you declare the actor with `@ray.remote` (without passing in `max_reconstructions`), then Ray won't try to reconstruct the actor. Is the issue that you want the actor to be reconstructed if it fails, but after it exits intentionally you don't want it to be reconstructed? In that case, then adding in something like `ray.exit_actor()` seems fine.

Reply:
@robertnishihara Yes, we need to pass `max_reconstructions` because we need the FO mechanism to keep the actor alive. However, at the same time, we need to shutdown this actor...
OK, then we can add `ray.exit_actor()`. In this case, the test may not be necessary, but the separated lock will be kept, right?

Reply:
@robertnishihara exactly, that's our scenario. 
@guoyuhong yep, we should still separate locks. That would benefit performance and prevent potential dead locks.